487 - IdsSwappable / swaplist

jmacaluso711 opened the pull request:
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR adds the IdsSwap list as well as IdsSwappable component. The abstracted component is currently named `ids-swappable` which can be used to contain any component you would like to with drag and drop.

**Related github/jira issue (required)**:
closes #487 

**Steps necessary to review your pull request (required)**:
- pull and run branch
- go to http://localhost:4300/ids-swaplist
- make sure you can select an item and drag it to another list.
- test multiple selection
- test select and swap with arrow buttons
- Reference enterprise: https://main-enterprise.demo.design.infor.com/components/swaplist/example-index.html
- check example usage of `ids-swappable` with `ids-cards` at http://localhost:4300/ids-swaplist/cards.html (WIP) missing come styling.

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

<!-- After submitting your PR, please check back to make sure checks on the PR -->

tmcconechy commented on src/components/ids-swappable/ids-swappable.d.ts:
```diff
@@ -0,0 +1,12 @@
+
+export default class IdsSwappable extends HTMLElement {
+  /* Sets the `multi-select` property to enable/disable multi-select ids-swappable-items */
+  multiSelect?: boolean;
```
On some other components we have `selection="multiple"` so maybe a slight inconsistencY

tmcconechy commented on src/components/ids-swappable/ids-swappable.js:
```diff
@@ -0,0 +1,246 @@
+import styles from './ids-swappable.scss';
+
+/**
+ * IDS SwapList Component
```
```suggestion
 * IDS Swappable Component
```

tmcconechy commented on src/components/ids-swappable/ids-swappable.js:
```diff
@@ -0,0 +1,246 @@
+
+  static get attributes() {
+    return [
+      attributes.ACTIVE,
```
Do we need ` ...super.attributes,`?

EdwardCoyle replied:
@tmcconechy there's a handful of mixins applied in `ids-swappable-base.js` so I'd say "yes" here

tmcconechy commented on src/components/ids-swappable/ids-swappable.js:
```diff
@@ -0,0 +1,246 @@
+   * @readonly
+   * @memberof IdsSwappable
+   */
+  get overEl() {
```
Maybe we make this overElement? (spelled out vs el?)

tmcconechy reviewed (changes requested):
Looks great a couple smaller comments. Also on the examples:

https://main-enterprise.demo.design.infor.com/components/swaplist/example-search.html
- compared to http://localhost:4300/ids-swaplist there is a 1px gap between the items 
- the placeholder is grey/slate/black and using dashes vs blue dotted

Biggest two questions are:
- removal of the template
- multiselect vs selection in other places...

Thanks!

danielortiz1982 commented on src/components/ids-swappable/ids-swappable-item.js:
```diff
@@ -0,0 +1,278 @@
+   * Handle the drag events
+   */
+  #handleDragEvents() {
+    this.removeEventListener('dragstart', this.#dragStart.bind(this));
```
If supported, I think we should use `onEvent` for all of these event handlers. If it's not supported I think it's fine for now.

tmcconechy replied:
yeah good catch - as onEvent will kill the event listener on delete

jmacaluso711 replied:
I tried using onEvent for these but getting this console error: `target.addEventListener is not a function` not exactly sure what the issue it.

tmcconechy replied:
There is another signature for onEvent..Did you get the second param?

```
this.onEvent('click', this, () => this.#dragStart());
```

EdwardCoyle commented on src/components/ids-swappable/ids-swappable-item.scss:
```diff
@@ -0,0 +1,40 @@
+@import '../../core/ids-base';
+
```
@jmacaluso711 I'll leave this one to your discretion because I'm not 100% sure if we need this...

If there is any chance at all that these styles might need to live in a standalone-CSS environment, we should try to figure out how to use them in CSS classes that would be applied to a shadow root container element instead.  However, I can see it's currently just a slot, so that would maybe change the design of this... 

I'm thinking it may not be necessary due to all of this just facilitating the swap behavior, but strikes me as something that may come up in the future.

EdwardCoyle reviewed (commented):
This is very nicely done @jmacaluso711!

I had one comment on functionality here that is minor and maybe ok to leave for a Phase 2:
- Open http://localhost:4300/ids-swaplist
- Highlight one or multiple items in List #0
- Click the "Swap Item Right" button in List #1
- See the highlighted item(s) jump from List #0 to List #2
- The old swaplist didn't do this, but we could potentially fix this by not allowing swapping via buttons unless items within those buttons' list are highlighted (disable the buttons otherwise?)

One other thing that may be unrelated to your changes... the two buttons in List #1 both say "Swap Item Left" in their text content (accessibility problem)

Just had another comment below that is also something we can push to later

jmacaluso711 commented:
@tmcconechy @danielortiz1982 @EdwardCoyle - Ready for re-review.. I think I addressed all comments, let me know if I missed anything.

@EdwardCoyle I ended up finding an easy fix for the swap button bug you found

tmcconechy reviewed (commented):
@jmacaluso711 the only thing im just noticing after reviewing the https://percy.io/00aaa841/enterprise-wc/builds/16355875/changed/920684642?browser=chrome&browser_ids=20&subcategories=unreviewed%2Cchanges_requested&viewLayout=side-by-side&viewMode=new&width=1280&widths=375%2C1280 (not as obvious on http://localhost:4300/ids-swaplist)

Is that there is an un-needed scrollbar?

tmcconechy commented:
@jmacaluso711  i synced this with main for you after the big merge. Looks good to me but you might want to take a quick look everything is as expected.

One thing i wondered about is

- http://localhost:4300/ids-swaplist/cards.html -> seems like the cards are too big and the select state looks a little strange. I didnt notice this the first time i reviewed. you could use this smaller actionable card maybe for the example https://main.wc.design.infor.com/ids-card/
- should this example be better filled under ids-swappable ?

jmacaluso711 commented:
@tmcconechy Yeah I was trying to show that you can put anything inside an `ids-swappable-item`, but didn't spend time making the example look right. I will move the demo to `ids-swappable` and simplify, maybe just use `ids-text` for now

jmacaluso711 commented:
@tmcconechy @EdwardCoyle ready for review

EdwardCoyle reviewed (approved):
Looks great @jmacaluso711
